### PR TITLE
CB-20679 Periscope can end up executing a scale operation against the

### PR DESCRIPTION
**Pull request by sidseth**
wrong cluster (Load Based Scaling).

Changes the various LoadBasedAdjustmentServices to not use instance variables, which otherwise get clobbered by multiple threads dealing with different clusters.

See detailed description in the commit message.

### CODE REVIEW

**Review by aakashshankar — APPROVED** [review]

File: autoscale/src/main/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentService.java [hunk 1]
```diff
@@ -82,28 +67,29 @@ public void pollYarnMetricsAndScaleCluster(Cluster cluster, String pollingUserCr
...
                 .filter(mandatoryDownscale -> mandatoryDownscale < 0).map(downscale -> -1 * downscale);
 
         YarnScalingServiceV1Response yarnResponse =
-                yarnMetricsClient.getYarnMetricsForCluster(cluster, stackV4Response, policyHostGroup, pollingUserCrn, maxDecommissionNodeCount);
+                yarnMetricsClient.getYarnMetricsForCluster(cluster, stackV4Response, policyHostGroupInstanceInfo.getPolicyHostGroup(),
+                        pollingUserCrn, maxDecommissionNodeCount);
 
         if (cluster.getUpdateFailedDetails() != null && pollingUserCrn.equals(cluster.getMachineUserCrn())) {
             // Successful YARN API call
             clusterService.setUpdateFailedDetails(cluster.getId(), null);
```
Comment (biharitomi) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentService.java:82]: I know it is not touched now, but can we have logs here what was the details of the failure?
Why does it call setUpdateFailedDetails with `null` details when based on the condition there should be details?

**Review by biharitomi — CHANGES_REQUESTED** [review]

**Comment by biharitomi** [comment]
I pasted few questions/requests, but what is a said fact that this "DTO" introduction change in multiple classes doesn't have any impact on any unit tests. :worried:

**Comment by lajosrodek** [comment]
To @biharitomi 's last note: I agree that the lack of UT coverage of this specific part of the code is not desirable. Due to the company-wide initiative of striving for improving system & unit test coverage of the codebase, I would vote for adding UT for these parts of `autoscale` as well. I anticipate this may be a legacy logic, and not all kinds of refactors demand an explicit must for creating UT (e.g. simple class renames), but the changes here are simply more complex than that.

@sidseth Could you please add some UT at least for the affected functionality? :smile:

File: autoscale/src/main/java/com/sequenceiq/periscope/service/LoadAlertPolicyHostGroupInstanceInfo.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    public LoadAlertPolicyHostGroupInstanceInfo(LoadAlert loadAlert, String policyHostGroup, Map<String, String> hostFqdnsToInstanceId,
+            List<String> servicesHealthyInstanceIds, List<String> stoppedHostInstanceIds) {
+        this.loadAlert = loadAlert;
+        this.policyHostGroup = policyHostGroup;
+        this.hostFqdnsToInstanceId = hostFqdnsToInstanceId;
+        this.servicesHealthyInstanceIds = servicesHealthyInstanceIds;
+        this.stoppedHostInstanceIds = stoppedHostInstanceIds;
+    }
+
+    public static LoadAlertPolicyHostGroupInstanceInfo create(StackResponseUtils stackResponseUtils, StackV4Response stackV4Response, LoadAlert loadAlert) {
```
[the remark this answers is not in the document]
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/LoadAlertPolicyHostGroupInstanceInfo.java]: Removed this method

File: autoscale/src/main/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentService.java (hunk 1 above)
Comment (aakashshankar) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentService.java:88]: Added a log line, this block is meant to reset updateFailedDetails in case of a successful API call made wth machineUserCrn.

**Comment by aakashshankar** [comment]
@lajosrodek @biharitomi , yes the coverage needs to be improved in these aspects
However, the rationale behind this fix was to address the multithreaded nature of periscope evaluators and accessing certain instance variables which initially caused discrepancies when issuing scaling requests to CB. The classes where this change has been introduced `RegularScalingAdjustmentService`, `StopStartScalingAdjustmentService`, `YarnBasedScalingAdjustmentService` has UTs which test upscales/downscales in different scenarios

**Comment by sidseth** [comment]
> To @biharitomi 's last note: I agree that the lack of UT coverage of this specific part of the code is not desirable. Due to the company-wide initiative of striving for improving system & unit test coverage of the codebase, I would vote for adding UT for these parts of `autoscale` as well. I anticipate this may be a legacy logic, and not all kinds of refactors demand an explicit must for creating UT (e.g. simple class renames), but the changes here are simply more complex than that.
> 
> @sidseth Could you please add some UT at least for the affected functionality? 😄

@lajosrodek , @biharitomi - as @aakashshankar mentioned elsewhere - this change deals with fixing some issues which show up in a multi-threaded setup. There is no functional change here. The individual classes already have quite a few unit tests in place. (Please point out where the UT coverage is lacking for this part of the code. UT coverage is absolutely critical - more than any subsequent testing; easiest/most reliable way to catch corner cases) A unit test to exercise the multi-threading to simulate this bug will not be very effective. Using instance variables should've been caught by me in the review before this went in.

Also - given this affects production, this needs to go in, and I'm not sure this particular PR is the place to bring up concerns about other parts of the code. How does it matter if there's a static method instead of a constructor?

File: autoscale/src/main/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentService.java (hunk 1 above)
Comment (biharitomi) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/YarnBasedScalingAdjustmentService.java:88]: Thanks, now it is more clear.

**Comment by lajosrodek** [comment]
@sidseth @aakashshankar  Understood, thanks for the clarification. I am fine with looking aside the UT side; testing multi-threaded stuff is indeed not straightforward. That is why I did not "ask for changes", only added an innocent comment. :smile:

**Review by biharitomi — APPROVED** [review]

**Comment by biharitomi** [comment]
Agreed @sidseth @aakashshankar  The new "DTO" object hasn't used in public functions so there was now schema change that would break any unit tests.

**Comment by sidseth** [comment]
recheck
